File: topics/graphs/medium/0133-clone-graph/solution.py
# 0133 · Clone Graph
# Approach: DFS with hash map | Time: O(V+E) | Space: O(V)

from typing import Optional


class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node:
            return None
        cloned = {}

        def dfs(n):
            if n in cloned:
                return cloned[n]
            copy = Node(n.val)
            cloned[n] = copy
            for neighbor in n.neighbors:
                copy.neighbors.append(dfs(neighbor))
            return copy

        return dfs(node)

File: topics/graphs/medium/0133-clone-graph/solution.cpp
// 0133 · Clone Graph
// Approach: DFS with hash map | Time: O(V+E) | Space: O(V)

#include <vector>
#include <unordered_map>
using namespace std;

class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node(int v) : val(v) {}
};

class Solution {
    unordered_map<Node*, Node*> cloned;

    Node* dfs(Node* node) {
        if (!node) return nullptr;
        if (cloned.count(node)) return cloned[node];
        Node* copy = new Node(node->val);
        cloned[node] = copy;
        for (Node* neighbor : node->neighbors)
            copy->neighbors.push_back(dfs(neighbor));
        return copy;
    }
public:
    Node* cloneGraph(Node* node) {
        return dfs(node);
    }
};
